web/packages/teleterm/src/ui/DocumentGatewayApp/AppGateway.tsx

```diff
@@ -94,6 +101,53 @@ export function AppGateway(props: {
     gateway.protocol === 'TCP' && gateway.targetSubresourceName;
   const targetPortRef = useRef<HTMLInputElement>(null);
 
+  const [tcpPortsAttempt, getTcpPorts] = useAsync(
+    useCallback(
+      () =>
+        retryWithRelogin(ctx, targetUri, () =>
+          tshd
+            .getApp({ appUri: targetUri })
+            .then(({ response }) => response.app.tcpPorts)
+        ),
+      [ctx, targetUri, tshd]
+    )
+  );
+  const currentTargetPort = parseInt(gateway.targetSubresourceName);
+  const getTcpPortsForMenuLogin: () => Promise<LoginItem[]> =
+    useCallback(async () => {
+      const [tcpPorts, error] = await getTcpPorts();
+
+      if (error) {
+        throw error;
+      }
+
+      return tcpPorts
+        .filter(
+          portRange =>
+            // Filter out single-port port ranges that are equal to the current port.
+            portRange.endPort !== 0 || portRange.port != currentTargetPort
```

Comment:
im sure youve written/read the word `port` so many times its lost all its meaning. just in these few things im like 🥴

Reply:
When I was working on the earlier PR which added `targetPort` to `OfflineGateway`, I started using `targetPort` everywhere, even in places where I really meant `localPort`. 🥲

> Thanks for making the commits easy to review!

Np, shuffling commits around and splitting them is sooooooooo much easier with jj.